File: LeetCodeProblemsGo/LeetCode1458MaxDotProductOfTwoSubsequences/LeetCode1458MaxDotProductOfTwoSubsequences.go
package LeetCode1458MaxDotProductOfTwoSubsequences

import "math"

func calcDp(dp [][]int, nums1, nums2 []int, length1, length2, firstStart, secondStart int) int {
	if firstStart >= length1 {
		return math.MinInt32
	}
	if secondStart >= length2 {
		return math.MinInt32
	}
	if dp[firstStart][secondStart] != math.MinInt32 {
		return dp[firstStart][secondStart]
	}
	if firstStart == length1-1 && secondStart == length2-1 {
		dp[firstStart][secondStart] = nums1[firstStart] * nums2[secondStart]
		return dp[firstStart][secondStart]
	}
	delta := nums1[firstStart] * nums2[secondStart]
	result := delta
	if num := calcDp(dp, nums1, nums2, length1, length2, firstStart+1, secondStart+1); num > 0 {
		result = delta + num
	}

	result = max(result, calcDp(dp, nums1, nums2, length1, length2, firstStart, secondStart+1))
	result = max(result, calcDp(dp, nums1, nums2, length1, length2, firstStart+1, secondStart))
	dp[firstStart][secondStart] = result
	return result
}
func maxDotProduct(nums1 []int, nums2 []int) int {
	maxProduct := math.MinInt32
	length1 := len(nums1)
	length2 := len(nums2)
	maxLength := max(length1, length2)
	dp := make([][]int, maxLength)
	for i := 0; i < length1; i++ {
		dp[i] = make([]int, length2)
		for j := 0; j < length2; j++ {
			dp[i][j] = math.MinInt32

		}
	}
	for i := 0; i < length1; i++ {
		for j := i; j < length2; j++ {

			maxProduct = max(calcDp(dp, nums1, nums2, length1, length2, i, j), maxProduct)

		}
	}
	return maxProduct
}

File: LeetCodeProblemsCMake/Problems/LeetCode1458MaxDotProductOfTwoSubsequences.cpp
//
// Created by saeed on 1/30/26.
//

#include "LeetCode1458MaxDotProductOfTwoSubsequences.h"

#include <stdint.h>
using namespace std;
int calcDp(vector<int>& nums1,vector<int>& nums2, vector<vector<int>>& dp,int length1,int length2,int first,int second) {
    if (first>=length1) {
        return INT32_MIN;
    }
    if (second>=length2) {
        return INT32_MIN;
    }
    if (dp[first][second]!=INT32_MIN) {
        return dp[first][second];
    }
    dp[first][second]= nums1[first]*nums2[second];
    int next=calcDp(nums1,nums2,dp,length1,length2,first+1,second+1);
    if (next>0) {
        dp[first][second]+=next;
    }
    dp[first][second]=max(dp[first][second],calcDp(nums1,nums2,dp,length1,length2,first+1,second));
    dp[first][second]=max(dp[first][second],calcDp(nums1,nums2,dp,length1,length2,first,second+1));
    return dp[first][second];
}
int LeetCode1458MaxDotProductOfTwoSubsequences::maxDotProduct(vector<int> &nums1, vector<int> &nums2) {
    int length1 =static_cast<int>(  nums1.size());
    int length2 = static_cast<int>(nums2.size());

    auto dp=vector<vector<int>>(length1, vector<int>(length2, INT32_MIN));
    int result= INT32_MIN;
    for (int i=0;i<length1;i++) {
        for (int j=0;j<length2;j++) {
            result=max(result,calcDp(nums1,nums2,dp,length1,length2,i,j) );
        }
    }
    return result;
}
